src/pyqasm/preprocess.py

```diff
@@ -0,0 +1,203 @@
+    result = "\n".join(ctx.base_file_header) + "\n\n" + "\n\n".join(ctx.results)
+    return result
+
+
+def _get_custom_includes(
+    ctx: IncludeContext, current_file: str, line: str, line_idx: int
+) -> list[tuple[str, int, int]]:
+    """Extracts the custom include file name from a line.
+
+    Args:
+        line (str): The line containing the include statement.
+
+    Returns:
+        list[tuple[str, int, int]]: A list of tuples containing the include filename,
+            line number, and column number where the include was found.
+    """
+    includes = []
+    # Search for custom include statements
+    match = PATTERNS["include_custom"].match(line)
+    if match:
+        include_filename = match.group(1)
+        # Ignore circular imports
+        if include_filename.strip() == current_file.strip():
+            col = line.index(include_filename) + 1
+            raise ValidationError(
+                f"Circular include detected for file '{include_filename}'"
+                f" at line {line_idx + 1}, column {col}: '{line.strip()}'"
+            )
+        # New include file to process found
+        if include_filename not in STD_FILES and include_filename not in ctx.visited:
+            col = line.index(include_filename) + 1
+            includes.append((include_filename, line_idx + 1, col))
+    return includes
+
+
+def _resolve_include_path(base_file: str, file_to_include: str) -> str | None:
+    """Resolve the include path for a given file.
+
+    Args:
+        base_file (str): The base file from which the include is being made.
+        file_to_include (str): The file to include.
+
+    Returns:
+        str | None: The resolved include path, or None if not found.
+    """
+    possible_paths = [os.path.join(os.path.dirname(base_file), file_to_include), file_to_include]
+    for path in possible_paths:
+        if os.path.isfile(path):
+            return path
+    return None
+
+
+def _collect_headers(ctx: IncludeContext, base_filename: str) -> None:
+    """Collects the header lines (OPENQASM and standard includes) from the base file.
+    Args:
+        base_filename (str): The base filename to read.
+    Returns:
+        None: Modifies the context in place.
+    """
+
+    with open(base_filename, "r", encoding="utf-8") as f:
+        program = f.read()
+
+    for line in program.splitlines():
+
+        if PATTERNS["openqasm"].match(line) or PATTERNS["include_standard"].match(line):
+            if PATTERNS["openqasm"].match(line) and line.strip() not in ctx.base_file_header:
+                # ensure openqasm line comes first
+                ctx.base_file_header.insert(0, line.strip())
+            if line.strip() not in ctx.base_file_header:
+                ctx.base_file_header.append(line.strip())
+            if PATTERNS["include_stdgates"].match(line):
+                ctx.include_stdgates = True
+            if PATTERNS["include_qelib1"].match(line):
+                ctx.include_qelib1 = True
+
```

Comment:
Since `OPENQASM` header should be the first line of the program and we only need to account for the existence standard gates' files in code, we can `break` the loop once all these criteria are met i.e. - 
```python
if ctx.include_stdgates and ctx.include_qelib1 and len(ctx.ctx.base_file_header) > 0 : 
     break 
```

Just make sure to double check the length condition with some tests